lib/integrations/azureoidc/accessgraph_sync.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package azureoidc
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/cloud/provisioning"
+	"github.com/gravitational/teleport/lib/config"
+	"github.com/gravitational/teleport/lib/msgraph"
+	tslices "github.com/gravitational/teleport/lib/utils/slices"
+)
+
+// graphAppId is the pre-defined application ID of the Graph API
+// Ref: [https://learn.microsoft.com/en-us/troubleshoot/entra/entra-id/governance/verify-first-party-apps-sign-in#application-ids-of-commonly-used-microsoft-applications].
+const graphAppId = "00000003-0000-0000-c000-000000000000"
+
+var requiredGraphRoleNames = []string{
+	"User.ReadBasic.All",
+	"Group.Read.All",
+	"Directory.Read.All",
+	"User.Read.All",
+	"Policy.Read.All",
+}
+
+func newManagedIdAction(cred *azidentity.DefaultAzureCredential, subId string, managedId string, roleName string) (*provisioning.Action, error) {
+	runnerFn := func(ctx context.Context) error {
+		// Create the role
+		roleDefCli, err := armauthorization.NewRoleDefinitionsClient(cred, nil)
+		if err != nil {
+			return trace.Wrap(fmt.Errorf("failed to create role definitions client: %w", err))
```

Comment:
```suggestion
			return trace.BadParameter("failed to create role definitions client: %v", err)
```

Same for all the others